lib/bolt/executor.rb

```diff
@@ -373,83 +382,156 @@ def run_plan(scope, plan, params)
+          # Stop any spins that might have started from a previous action so we
+          # don't get multiple spinners.
+          publish_event(type: :stop_spin)
+          r = fiber.resume
+        # If the Fiber is dead that means it's returned a result, which we
+        # automatically store on the Bolt::Future object when we resume it
+        # (since it can't be retrieved once the Fiber dies).
+        else
+          # Reset the sleep since something successfully resumed
+          @sleep = 0.1
+          @logger.trace("Retrieving result from future '#{fiber.name}'") if in_parallel
+          # Get the result from the Bolt::Future object
+          # Remove the fiber
+          if fiber == main_plan
+            fiber_results.main_plan = fiber.value
+            plan_fibers.main_plan = nil
           else
-            raise e
+            fiber_results.fibers[fiber.name] = fiber.value
+            plan_fibers.fibers.delete(fiber)
           end
         end
       end
 
-      Bolt::Yarn.new(fiber, index)
-    end
+      return unless r.is_a?(Symbol) && r == :unfinished
+      @logger.trace("Nothing can be resumed. Rechecking in #{@sleep} seconds.") if in_parallel
 
-    def handle_event(event)
-      case event[:type]
-      when :node_result
-        @thread_completed = true
-      end
+      # Rather than trying
+      @sleep = @sleep > 1 ? @sleep : @sleep * 2
+      sleep(@sleep)
     end
 
-    def round_robin(skein)
-      subscribe(self, [:node_result])
-      results = Array.new(skein.length)
-      @in_parallel = true
-      publish_event(type: :stop_spin)
+    def plan_complete?
+      plan_fibers.fibers.empty? && main_plan.nil?
+    end
 
-      until skein.empty?
-        @thread_completed = false
-        r = nil
+    def wait(fibers, timeout, catch_errors: false, **_kwargs)
+      names = fibers.map(&:name)
+      @logger.trace("Waiting for futures #{names} to complete.")
 
-        skein.each do |yarn|
-          if yarn.alive?
-            publish_event(type: :stop_spin)
-            r = yarn.resume
-          else
-            results[yarn.index] = yarn.value
-            skein.delete(yarn)
-          end
+      # Wait until all the Fibers have completed
+      start = wait_now
+      until names.all? { |n| fiber_results.fibers.keys.include?(n) }
+        if timeout && (wait_now - start).to_i >= timeout
+          # If a Fiber returned a value it will already be deleted from
+          # plan_fibers, if not delete it now so it does not continue
+          # executing.
+          plan_fibers.fibers.delete_if { |f| fibers.include?(f) }
+          break
         end
+        round_robin(plan_fibers.fibers)
+      end
 
-        next unless r == 'unfinished'
-        sleep(0.1) until @thread_completed || skein.empty?
+      # Collect and return results from the specific fibers waited on in order.
+      result = names.each_with_object([]) do |name, acc|
+        acc << if (f = fiber_results.fibers[name])
```

Comment:
This is an example where I don't think we need the `fiber_results` why not just ask the completed fibers what their `value` is?